frame/contracts/src/wasm/runtime.rs

```diff
@@ -485,25 +481,44 @@ impl<'a, E: Ext + 'a> Runtime<'a, E> {
 
 	/// Converts the sandbox result and the runtime state into the execution outcome.
 	pub fn to_execution_result(self, sandbox_result: Result<(), wasmi::Error>) -> ExecResult {
+		use wasmi::core::TrapCode::OutOfFuel;
 		use TrapReason::*;
+
 		match sandbox_result {
 			// Contract returned from main function -> no data was returned.
 			Ok(_) => Ok(ExecReturnValue { flags: ReturnFlags::empty(), data: Vec::new() }),
+			// `OutOfGas` when host asks engine to consume more than left in the _store_.
+			// We should never get this case, as gas meter is being charged (and hence raises error)
+			// first.
+			Err(wasmi::Error::Store(_)) => Err(Error::<E::T>::OutOfGas.into()),
 			// Contract either trapped or some host function aborted the execution.
 			Err(wasmi::Error::Trap(trap)) => {
+				if let Some(code) = trap.trap_code() {
+					match code {
+						// `OutOfGas` during engine execution.
+						OutOfFuel => return Err(Error::<E::T>::OutOfGas.into()),
+						_ => (),
+					}
+				}
 				// If we encoded a reason then it is some abort generated by a host function.
-				// Otherwise the trap came from the contract.
-				let reason: TrapReason = trap.downcast().ok_or(Error::<E::T>::ContractTrapped)?;
-				match reason {
-					Return(ReturnData { flags, data }) => {
-						let flags =
-							ReturnFlags::from_bits(flags).ok_or(Error::<E::T>::InvalidCallFlags)?;
-						Ok(ExecReturnValue { flags, data })
-					},
-					Termination =>
-						Ok(ExecReturnValue { flags: ReturnFlags::empty(), data: Vec::new() }),
-					SupervisorError(error) => return Err(error.into()),
+				if let Some(reason) = &trap.downcast_ref::<TrapReason>() {
+					match &reason {
+						Return(ReturnData { flags, data }) => {
+							let flags = ReturnFlags::from_bits(*flags)
+								.ok_or(Error::<E::T>::InvalidCallFlags)?;
+							return Ok(ExecReturnValue { flags, data: data.to_vec() })
+						},
+						// TODO check sanity
```

Comment:
needs to be resolved before merging?

Reply:
something looked weird to me here, I left it here not to forget. will be removed bf merge